core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java

```diff
@@ -827,6 +828,47 @@ static RelDataType deriveTypeSplit(SqlOperatorBinding operatorBinding,
           .withOperandTypeInference(InferTypes.RETURN_TYPE)
           .withKind(SqlKind.CONCAT2);
 
+  private static RelDataType namedStructReturnType(SqlOperatorBinding operatorBinding) {
+    if (operatorBinding.getOperandCount() < 2) {
+      throw operatorBinding.newError(Static.RESOURCE.namedStructRequiresTwoOrMoreArgs());
+    }
+    if (operatorBinding.getOperandCount() % 2 != 0) {
+      throw operatorBinding.newError(Static.RESOURCE.namedStructRequiresEvenNumberOfArgs());
+    }
+    List<String> keys = new ArrayList<String>();
+    List<RelDataType> values = new ArrayList<RelDataType>();
```

Comment:
calcite try best to using final,
final List<String> keys 
final List<RelDataType> values

Reply:
We should use Scala 🤣 .